src/Media_Command.php

```diff
@@ -388,4 +390,77 @@ private function needs_regeneration( $att_id, $fullsizepath ) {
 		}
 		return false;
 	}
+
+	// Like WP's get_intermediate_image_sizes(), but removes sizes that won't be generated for a particular attachment due to its being on or below their thresholds.
+	private function get_intermediate_image_sizes_for_attachment( $fullsizepath, $is_pdf, $metadata ) {
+
+		// Need to get width, height of attachment for image_resize_dimensions().
+		$editor = wp_get_image_editor( $fullsizepath );
+		if ( is_wp_error( $editor ) ) {
+			WP_CLI::warning( $editor->get_error_message() );
+			return array();
+		}
+		if ( is_wp_error( $result = $editor->load() ) ) {
+			WP_CLI::warning( $result->get_error_message() );
+			unset( $editor );
+			return array();
+		}
+		list( $width, $height ) = array_values( $editor->get_size() );
+		unset( $editor );
+
+		$sizes = array();
+		foreach ( $this->get_intermediate_sizes( $is_pdf, $metadata ) as $name => $size ) {
+			// Need to check destination and original width or height differ before calling image_resize_dimensions(), otherwise it will return non-false.
+			if ( ( $width !== $size['width'] || $height !== $size['height'] ) && image_resize_dimensions( $width, $height, $size['width'], $size['height'], $size['crop'] ) ) {
+				$sizes[] = $name;
+			}
+		}
+		return $sizes;
+	}
+
+	// Like WP's get_intermediate_image_sizes(), but returns associative array with name => size info entries (and caters for PDFs also).
+	private function get_intermediate_sizes( $is_pdf, $metadata ) {
+		if ( $is_pdf ) {
+			// Copied from wp_generate_attachment_metadata() in "wp-admin/includes/image.php".
+			$fallback_sizes = array(
+				'thumbnail',
+				'medium',
+				'large',
+			);
+			$intermediate_image_sizes = apply_filters( 'fallback_intermediate_image_sizes', $fallback_sizes, $metadata );
+		} else {
+			$intermediate_image_sizes = get_intermediate_image_sizes();
+		}
+
+		// Adapted from wp_generate_attachment_metadata() in "wp-admin/includes/image.php".
+
+		$_wp_additional_image_sizes = wp_get_additional_image_sizes();
+
+		$sizes = array();
+		foreach ( $intermediate_image_sizes as $s ) {
+			if ( isset( $_wp_additional_image_sizes[ $s ]['width'] ) ) {
+				$sizes[ $s ]['width'] = intval( $_wp_additional_image_sizes[ $s ]['width'] );
+			} else {
+				$sizes[ $s ]['width'] = intval( get_option( "{$s}_size_w" ) );
+			}
+
+			if ( isset( $_wp_additional_image_sizes[ $s ]['height'] ) ) {
+				$sizes[ $s ]['height'] = intval( $_wp_additional_image_sizes[ $s ]['height'] );
+			} else {
+				$sizes[ $s ]['height'] = intval( get_option( "{$s}_size_h" ) );
+			}
+
+			if ( isset( $_wp_additional_image_sizes[ $s ]['crop'] ) ) {
+				$sizes[ $s ]['crop'] = !! $_wp_additional_image_sizes[ $s ]['crop'];
+			} else {
+				// Force PDF thumbnails to be soft crops.
+				if ( $is_pdf && 'thumbnail' === $s ) {
+					$sizes[ $s ]['crop'] = false;
+				} else {
+					$sizes[ $s ]['crop'] = !! get_option( "{$s}_crop" );
```

Comment:
Is this a double negation? What's the reason for that? In case it is a shorthand to transform into a boolean, we should prefer a cast instead: `(bool) get_option()`.